contrib/opencensus-ext-azure/opencensus/ext/azure/common/exporter.py

```diff
@@ -0,0 +1,132 @@
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import atexit
+import threading
+import time
+
+from six.moves import queue
+from opencensus.ext.azure.common import Options
+
+
+class QueueEvent(object):
+    def __init__(self, name):
+        self.name = name
+        self.event = threading.Event()
+
+    def __repr__(self):
+        return ('{}({})'.format(type(self).__name__, self.name))
+
+    def set(self):
+        return self.event.set()
+
+    def wait(self, timeout=None):
+        return self.event.wait(timeout)
+
+
+class BaseExporter(object):
+    EXIT_EVENT = QueueEvent('EXIT')
+
+    def __init__(self, **options):
+        self.options = Options(**options)
+        self.grace_period = self.options.grace_period
+        self.interval = self.options.export_interval
+        self.max_batch_size = self.options.max_batch_size
+        self._stopping = False
+        self._queue = queue.Queue(maxsize=self.options.max_queue_size)
+        self._thread = threading.Thread(target=self._thread_entry)
+        self._thread.daemon = True
+        self._thread.start()
+        atexit.register(self._stop, self.grace_period)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self._stop(None)
+
+    def __gets(self, count, timeout):
+        start_time = time.time()
+        elapsed_time = 0
+        cnt = 0
+        while cnt < count:
+            try:
+                item = self._queue.get(block=False)
+                yield item
+                if isinstance(item, QueueEvent):
+                    return
+            except queue.Empty:
+                break
+            cnt += 1
+        while cnt < count:
+            wait_time = max(timeout - elapsed_time, 0)
+            try:
+                item = self._queue.get(block=True, timeout=wait_time)
+                yield item
+                if isinstance(item, QueueEvent):
+                    return
+            except queue.Empty:
+                break
+            cnt += 1
+            elapsed_time = time.time() - start_time
+
+    def _gets(self, count, timeout):
+        return tuple(self.__gets(count, timeout))
+
+    def _stop(self, timeout):
+        start_time = time.time()
+        wait_time = timeout
+        if self._thread.is_alive() and not self._stopping:
+            self._stopping = True
+            self._queue.put(self.EXIT_EVENT, block=True, timeout=wait_time)
+            elapsed_time = time.time() - start_time
+            wait_time = timeout and max(timeout - elapsed_time, 0)
+        self._thread.join(timeout=wait_time)
+        if self._thread.is_alive():
+            return
+        return time.time() - start_time  # time taken to flush
```

Comment:
Something to keep in mind for the future: this is a good use case for [`time.monotonic`](https://docs.python.org/3/library/time.html#time.monotonic), but it's not clear that it's worth it if we still have to fall back to `time.time` for py2 compatibility.